contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,473 @@
+
+Before starting the `ControllerPublishVolume` operation, the external-attacher should add these finalizers to these Kubernetes API objects:
+
+* To the `VolumeAttachment` so that when the object is deleted, the external-attacher has an opportunity to detach the volume first. External attacher removes this finalized once the volume is fully detached from the node.
+* To the `PersistentVolume` referenced by `VolumeAttachment` so the the PV cannot be deleted while the volume is attached. External attacher needs information from the PV to perform detach operation. The attacher will remove the finalizer once all `VolumeAttachment` objects that refer to the PV are deleted, i.e. the volume is detached from all nodes.
+
+If the operation completes successfully, the external-attacher will:
+
+1. Set `VolumeAttachment.Status.Attached` field to true to indicate the volume is attached.
+2. Update the `VolumeAttachment.Status.AttachmentMetadata` field with the contents of the returned `PublishVolumeInfo`.
+3. Clear the `VolumeAttachment.Status.AttachError` field.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.Attached` field to still false to indicate the volume is not attached.
+2. Set the `VolumeAttachment.Status.AttachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The external-attacher may implement it’s own error recovery strategy, and retry as long as conditions specified for attachment above are valid. It is strongly recommended that the external-attacher implement an exponential backoff strategy for retries.
+
+The detach operation will be triggered by the deletion of the `VolumeAttachment` Kubernetes API objects. Since the `VolumeAttachment` Kubernetes API object will have a finalizer added by the external-attacher, it will wait for confirmation from the external-attacher before deleting the object.
+
+Once all the following conditions are true, the external-attacher should call `ControllerUnpublishVolume` against the CSI volume driver to detach the volume from the specified node:
+1. A `VolumeAttachment` Kubernetes API object is marked for deletion: the value for the `VolumeAttachment.metadata.deletionTimestamp` field is set.
+
+If the operation completes successfully, the external-attacher will:
+1. Remove its finalizer from the list of finalizers on the `VolumeAttachment` object permitting the delete operation to continue.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.Attached` field remains true to indicate the volume is not yet detached.
+2. Set the `VolumeAttachment.Status.DetachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The new API object called `VolumeAttachment` will be defined as follows:
```

Comment:
Devil's advocate: did we rule out making this just be new fields on PV ?

Reply:
Mostly because it makes detaching tricker.

Reply:
did you make this case in teh doc?  I can't find it without re-reading teh whole thing